embassy-net/src/udp.rs

```diff
@@ -103,6 +103,34 @@ impl<'a> UdpSocket<'a> {
         })
     }
 
+    /// Wait until the socket becomes readable.
+    ///
+    /// A socket is readable when a packet has been received, or when there are queued packets in
+    /// the buffer.
+    pub async fn wait_recv_ready(&self) {
+        poll_fn(move |cx| self.poll_recv_ready(cx)).await
+    }
+
+    /// Wait until a datagram can be read.
+    ///
+    /// When no datagram is readable, this method will return `Poll::Pending` and
+    /// register the current task to be notified when a datagram is received.
+    ///
+    /// When a datagram is received, this method will return `Poll::Ready`.
+    pub fn poll_recv_ready(&self, cx: &mut Context<'_>) -> Poll<()> {
+        self.with_mut(|s, _| {
+            match s.peek() {
```

Comment:
you can use `.can_recv()` instead of actually peeking, which avoids having to use `unreachable!()`.

Reply:
@Dirbaio 

Are we sure that `can_recv()` is an OK replacement for `peek()`?

`peek()` also calls something named [dequeue_adding](https://github.com/smoltcp-rs/smoltcp/blob/a2b92ed6e8d7ae7b1229665c5b3f2e801de111e9/src/storage/packet_buffer.rs#L166), while `can_recv()` doesn't.

Using `can_recv`, can we end up in a situation where `recv_ready` becomes "ready" - yet - a following call to `read`/`recv` still pends, because the only thing in the ringbuffer was padding (whatever "padding" is)?